design-proposals/vm-memory-snapshot.md

```diff
@@ -0,0 +1,151 @@
+# Overview
+
+Proposal for taking VM RAM snapshot which can be either a stand alone action, or alongside Online VMSnapshot.
+
+## Motivation
+One of the methods of troubleshooting guest virtual machines is getting a memory dump and analyzing it. As this is a standard capability in most VM management systems, it is also required for kubevirt’s VMs.
+
+## Goals
+Have a mechanism to take and get a VM memory snapshot.
+This mechanism can be either during the Online VMSnapshot process to get a complete snapshot of a VM: disks and Ram alike, or as a stand alone command to get the memory dump only.
+
+## Non Goals
+Currently the memory snapshot will not be used for VM hibernation, not in the VM restore process.
+
+## User Stories
+As a Kubeviert user I would like to take a snapshot of a running VM so that I can later inspect its disk and memory
+A an Kubevirt user I would like to get a memory dump of a running VM so that I can later inspect its memory.
+
+## Repos
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+As mentioned the memory snapshot can be taken either in the Online VMSnapshot or as a stand alone command.
+To trigger the memory snapshot during the Online VMSnapshot one will add a parameter to the VMSnapshot yaml.
+To trigger a memory snapshot as a stand alone command a virtctl memoryDump command will be added.
+
+** Before doing either option a pvc should be mounted to the virt-launcher pod. A volume source type will be added to indicate that this is a volume without a matching disk since the VMI doesn't need to access the memory dump, only the virt launcher. This volume source will be able to be added to the VM spec as a volume or by using the Hot-plug API that will be extended to adding this volume without a disk.
+Only one memoryDump volume will be for the VM, the content will be overwritten in every memory dump call.
+
+## API
+
+### New volume source
```

Comment:
I think memoryDump best makes most sense as an imperative operation.  You call the subresource API, the operation is executed as quickly as possible, and when the call returns you know whether it succeeded or not.  That all requires the target of the dump be mounted in the virt launcher pod.  Do you agree?

If the dump target is not mounted like the rest of the other PVC based volumes (static or hotplug).  How would it be configured?  A new resource that references the VM/VMI?

Reply:
@rmohr In order for the dump to be included in the snapshot it made sense that the pvc that will contain the dump will be a part of the VM specifically some kind of volume, the pvc is attached to the virt-launcher pod like a volume it just not being mounted to the vmi as a disk also. Also for hibernation the pvc need to be a part of the vm even after it is stopped, so it seems like it should to be as part of the vm.
@mhenriks the operation will not return after the completion of the operation since from what I understand if it possible that the dump will be big and will take some time, so the plan is to send the command and then wait till the status of the dump volume will be updated to completed the operation.